sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala

```diff
@@ -224,6 +225,56 @@ case class Limit(limit: Int, child: SparkPlan)
 
 /**
  * :: DeveloperApi ::
+ * Take the first limit elements. and the limit can be any number less than Integer.MAX_VALUE.
+ * If it is terminal and is invoked using executeCollect, it probably cause OOM if the
+ * records number is large enough. Not like the Limit clause, this operator will not change
+ * any partitions of its child operator.
+ */
+@DeveloperApi
+case class LargeLimit(limit: Int, child: SparkPlan)
+  extends UnaryNode {
+  /** We must copy rows when sort based shuffle is on */
+  private def sortBasedShuffleOn = SparkEnv.get.shuffleManager.isInstanceOf[SortShuffleManager]
+
+  override def output: Seq[Attribute] = child.output
+
+  override def executeCollect(): Array[Row] = child.executeTake(limit)
+
+  protected override def doExecute(): RDD[InternalRow] = {
+    val rdd = if (sortBasedShuffleOn) {
+      child.execute().map(_.copy()).persist(StorageLevel.MEMORY_AND_DISK)
+    } else {
+      child.execute().persist(StorageLevel.MEMORY_AND_DISK)
```

Comment:
That's the main reason I put the WIP on the title, persist the RDD would be helpful to avoid the re-computation, but I have no idea how to release it unpersist it automatically.